src/build.cc

```diff
@@ -474,6 +486,133 @@ bool Plan::CleanNode(DependencyScan* scan, Node* node, string* err) {
   return true;
 }
 
+namespace {
+bool EdgeCom(const Edge* lhs, const Edge* rhs) {
+  // Note: > to sort in decreasing order.
+  return lhs->critical_time() > rhs->critical_time();
+}
+}  // namespace
+
+void Plan::ComputePriorityList(BuildLog* build_log) {
+
+  for (vector<Node*>::iterator it=targets_.begin();it!=targets_.end();++it){
+      //remove duplicate targets
+      targets_.resize(find(it+1,targets_.end(),*it)-targets_.begin());
+  }
+
+  vector<Edge*> edges;
+  for (map<Edge*, bool>::iterator it = want_.begin(), end = want_.end();
+       it != end; ++it) {
+    if (!it->second)
+      continue;
+    Edge* e = it->first;
+    edges.push_back(e);
+  }
+  
+
+  // Critical path scheduling.
+  // 0. Assign costs to all edges, using:
+  // a) The time the edge needed last time, if available.
+  // b) The average time this edge type needed, if this edge hasn't run before.
+  // c) A fixed cost if this type of edge hasn't run before.
+  for (vector<Edge*>::iterator it = edges.begin(), end = edges.end(); it != end;
+       ++it) {
+    Edge* edge = *it;
+    if (edge->is_phony())
+      continue;  
+    BuildLog::LogEntry* entry =
+        build_log->LookupByOutput(edge->outputs_[0]->path());
+    if (!entry) {
+      edge->run_time_ms_ = 1;
+      continue;
+    }
+
+    int duration = entry->end_time - entry->start_time;  // XXX: + 1?
+    edge->run_time_ms_ = duration;
+    // XXX: also count on rule, and globally
+  }
+
+
+  // Dump graph to stdout for debugging / prototyping.
+  if (false) {
+    for (vector<Edge*>::iterator it = edges.begin(), end = edges.end();
+         it != end; ++it) {
+      Edge* edge = *it;
+      Node* input = edge->inputs_[0];
+      Node* output = edge->outputs_[0];
+      printf("%s %s %d\n", input->path().c_str(), output->path().c_str(),
+             edge->run_time_ms_);
+    }
+  }
+
+
+  uint64_t prev_top_crit=0;
+
+  for (vector<Node*>::reverse_iterator tit = targets_.rbegin(),
```

Comment:
can you improve order of this algorithm?
It looks like O(|targets| * |build edge|), it can be long.

Reply:
for my inital idea, earlier built target should uses previous pass max critical time as inital value, so it is O(|targets| * |build edge|)  but if you do not care to save bits of type of critical time(uint64_t) very much , we can use the longest path length (suppose  N ) of whole graph as lowest priority target max critical time,  and assign all targets initial critical time as 
pow(2,M) N......  4N  2N  N 0
so we can just walk whole graph once to get N, then just follow nico's previous implementation, mark all critical time for all edges at once.  it may waste more bits than O(|targets| * |build edge|) , but I think which is also desirable.

this is a quick implementation which improve my own build process immediately,  useful when the longest job build time much more than all other small jobs.  in my own test (3500 sources),  total saved build time (8 min )is much more than this algorithms spent( which I havn't measured this exactly, at least I havn't seen big latency before first job is starting build) . 

to my suprise, without explicit targets specified, this critical priority list didn't show big win over default ninja(just as nico mentioned),  objects on ciritical path do build much earlier than default, but the link step, didn't being picked earlier very much. so I think the priority logic may still have something  missed.